File: apps/mobile_customer/test/models/bundle_b_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_customer/shared/models/rows.dart';

void main() {
  group('Bundle B - Deposit & Policy (V24)', () {
    test('BookingRow parses deposit & cancel fields', () {
      final json = {
        'id': 'b-bundle',
        'bookingCode': 'DKT-BUNDLE',
        'status': 'CONFIRMED',
        'currency': 'IDR',
        'subtotal': 100000,
        'total': 100000,
        'depositAmount': 25000,
        'depositRequired': true,
        'deposit_amount': 25000,
        'deposit_required': true,
        'cancelDeadline': '2026-09-09T10:00:00+07:00',
        'cancel_deadline': '2026-09-09T10:00:00+07:00',
        'rescheduleCount': 0,
        'reschedule_count': 0,
        'maxReschedule': 1,
        'max_reschedule': 1,
        'cancelPolicy': '24h_full_refund',
        'cancel_policy': '24h_full_refund',
        'version': 1,
        'startsAt': '2026-09-10T10:00:00+07:00',
        'endsAt': '2026-09-10T11:00:00+07:00',
        'createdAt': '2026-09-08T08:00:00+07:00',
      };
      final row = BookingRow.fromJson(json);
      expect(row.depositAmount, 25000);
      expect(row.depositRequired, true);
      expect(row.cancelDeadline, isNotNull);
      expect(row.rescheduleCount, 0);
      expect(row.maxReschedule, 1);
      expect(row.cancelPolicy, '24h_full_refund');
      expect(row.version, 1);
    });

    test('BookingRow defaults deposit false and reschedule 0/1', () {
      final row = BookingRow.fromJson({'id': 'b1'});
      expect(row.depositAmount, 0);
      expect(row.depositRequired, false);
      expect(row.rescheduleCount, 0);
      expect(row.maxReschedule, 1);
      expect(row.cancelPolicy, isNull);
      expect(row.cancelDeadline, isNull);
    });

    test('Cancel after deadline no refund logic (H-24)', () {
      final booking = BookingRow.fromJson({
        'id': 'b2',
        'startsAt': '2026-09-10T10:00:00+07:00',
        'cancelDeadline': '2026-09-09T10:00:00+07:00',
        'depositAmount': 20000,
        'cancelPolicy': '24h_full_refund',
      });
      final now = DateTime.parse('2026-09-09T11:00:00+07:00');
      final afterDeadline = booking.cancelDeadline != null && now.isAfter(booking.cancelDeadline!);
      expect(afterDeadline, true);
      // no refund flag
      final refund = afterDeadline ? 0 : booking.depositAmount;
      expect(refund, 0);
    });

    test('Reschedule limit 1 free then 409', () {
      final booking = BookingRow.fromJson({
        'id': 'b3',
        'rescheduleCount': 0,
        'maxReschedule': 1,
      });
      expect(booking.rescheduleCount < booking.maxReschedule, true);
      final afterFirst = BookingRow.fromJson({
        'id': 'b3',
        'rescheduleCount': 1,
        'maxReschedule': 1,
      });
      expect(afterFirst.rescheduleCount >= afterFirst.maxReschedule, true); // should 409
    });

    test('ServiceRow depositAmount parsed', () {
      final svc = ServiceRow.fromJson({'id': 's1', 'name': 'Haircut', 'price': 50000, 'duration': 30, 'depositAmount': 15000});
      expect(svc.depositAmount, 15000);
      final svc2 = ServiceRow.fromJson({'id': 's2', 'name': 'Cut'});
      expect(svc2.depositAmount, 0);
    });
  });

  group('Bundle B - Kalender sync (VCALENDAR & Google)', () {
    test('ICS generation format', () {
      // simulate ics generation logic used in backend
      String generateIcs(String bookingCode, DateTime start, DateTime end) {
        String fmt(DateTime d) {
          final utc = d.toUtc();
          String two(int n) => n.toString().padLeft(2, '0');
          return '${utc.year}${two(utc.month)}${two(utc.day)}T${two(utc.hour)}${two(utc.minute)}${two(utc.second)}Z';
        }
        return 'BEGIN:VCALENDAR\r\nVERSION:2.0\r\nBEGIN:VEVENT\r\nUID:${bookingCode}@dekat.id\r\nDTSTART:${fmt(start)}\r\nDTEND:${fmt(end)}\r\nSUMMARY:DEKAT Booking ${bookingCode}\r\nEND:VEVENT\r\nEND:VCALENDAR\r\n';
      }
      final start = DateTime.parse('2026-09-10T10:00:00+07:00');
      final end = DateTime.parse('2026-09-10T11:00:00+07:00');
      final ics = generateIcs('DKT-TEST1', start, end);
      expect(ics, contains('BEGIN:VCALENDAR'));
      expect(ics, contains('END:VCALENDAR'));
      expect(ics, contains('BEGIN:VEVENT'));
      expect(ics, contains('UID:DKT-TEST1@dekat.id'));
      expect(ics, contains('DTSTART:20260910T030000Z')); // 10+07 -> 03Z
      expect(ics, contains('DTEND:20260910T040000Z'));
    });

    test('Google Calendar link generation', () {
      String googleLink(String code, DateTime start, DateTime end) {
        String fmt(DateTime d) {
          final utc = d.toUtc();
          String two(int n) => n.toString().padLeft(2, '0');
          return '${utc.year}${two(utc.month)}${two(utc.day)}T${two(utc.hour)}${two(utc.minute)}${two(utc.second)}Z';
        }
        return 'https://calendar.google.com/calendar/render?action=TEMPLATE&text=DEKAT Booking ${Uri.encodeComponent(code)}&dates=${fmt(start)}/${fmt(end)}&details=Booking&location=DEKAT';
      }
      final start = DateTime.parse('2026-09-10T10:00:00+07:00');
      final end = DateTime.parse('2026-09-10T11:00:00+07:00');
      final url = googleLink('DKT-123', start, end);
      expect(url, startsWith('https://calendar.google.com'));
      expect(url, contains('DKT-123'));
      expect(url, contains('20260910T030000Z'));
    });
  });

  group('Bundle B - FAQ & Policy', () {
    test('FaqRow parses', () {
      final f = FaqRow.fromJson({'id': 'f1', 'question': 'Bagaimana cara booking?', 'answer': 'Pilih provider...', 'category': 'booking', 'sortOrder': 1, 'isActive': true});
      expect(f.question, contains('booking'));
      expect(f.category, 'booking');
      expect(f.isActive, true);
    });

    test('PolicyRow parses', () {
      final p = PolicyRow.fromJson({'id': 'p1', 'title': 'Kebijakan Pembatalan', 'body': '24h', 'type': 'cancellation', 'version': 1});
      expect(p.title, contains('Pembatalan'));
      expect(p.type, 'cancellation');
    });

    test('Policy cancellation 24h no refund after deadline', () {
      final policy = PolicyRow.fromJson({'id': 'p1', 'title': 'Kebijakan', 'body': 'Pembatalan sebelum 24 jam = refund penuh', 'type': 'cancellation'});
      expect(policy.type, 'cancellation');
      expect(policy.body, contains('24 jam'));
    });
  });
}

File: apps/web_public/src/lib/__tests__/bundleB.test.ts
import { publicApi } from '../api';

const mockFetch = vi.fn();
global.fetch = mockFetch as unknown as typeof fetch;

function mockResponse(data: unknown, ok = true, status = 200) {
  return { ok, status, json: () => Promise.resolve(data), text: () => Promise.resolve(typeof data === 'string' ? data : JSON.stringify(data)), headers: new Map([['content-type', 'application/json']]) };
}

describe('Bundle B - Deposit & Kalender & FAQ', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    localStorage.clear();
    localStorage.setItem('auth_user', JSON.stringify({ id: 'actor-1' }));
  });

  describe('deposit & pembatalan', () => {
    it('P: booking response includes deposit fields (V24)', async () => {
      const booking = { id: 'bk-1', bookingCode: 'DKT-TEST', status: 'CONFIRMED', depositAmount: 25000, depositRequired: true, cancelDeadline: '2026-09-09T10:00:00+07:00', rescheduleCount: 0, maxReschedule: 1, cancelPolicy: '24h_full_refund' };
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: booking }));
      const res = await publicApi.bookings.getById('bk-1');
      expect(res.data.depositAmount).toBe(25000);
      expect(res.data.depositRequired).toBe(true);
      expect(res.data.cancelPolicy).toBe('24h_full_refund');
    });

    it('N: cancel after deadline still returns 200 but no refund note in history (backend)', async () => {
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: { id: 'bk-1', status: 'CANCELLED' } }));
      const res = await publicApi.bookings.cancel('bk-1', 'late');
      expect(res.data.status).toBe('CANCELLED');
      const [, opts] = mockFetch.mock.calls[0];
      expect(opts.method).toBe('POST');
    });

    it('E: reschedule second time 409 Conflict', async () => {
      mockFetch.mockResolvedValueOnce(mockResponse({ success: false, message: 'Reschedule limit reached. Maximum 1 free reschedule(s) allowed' }, false, 409));
      await expect(publicApi.bookings.reschedule('bk-1', { newStartsAt: '2026-09-11T10:00:00+07:00', newEndsAt: '2026-09-11T11:00:00+07:00', expectedVersion: 1 })).rejects.toThrow();
    });

    it('P: reschedule first time succeeds', async () => {
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: { id: 'bk-1', status: 'CONFIRMED', rescheduleCount: 1 } }));
      const res = await publicApi.bookings.reschedule('bk-1', { newStartsAt: '2026-09-11T10:00:00+07:00', newEndsAt: '2026-09-11T11:00:00+07:00', expectedVersion: 1 });
      expect(res.data.rescheduleCount).toBe(1);
    });
  });

  describe('kalender sync', () => {
    it('P: ics endpoint returns text/calendar VCALENDAR', async () => {
      const ics = 'BEGIN:VCALENDAR\r\nVERSION:2.0\r\nBEGIN:VEVENT\r\nUID:test@dekat.id\r\nEND:VEVENT\r\nEND:VCALENDAR';
      mockFetch.mockResolvedValueOnce({ ok: true, status: 200, text: () => Promise.resolve(ics), json: () => Promise.resolve(ics) } as unknown as Response);
      const res = await publicApi.bookings.ics('bk-1');
      // apiClient.get with responseType text returns string in data
      expect(mockFetch).toHaveBeenCalled();
      const [url] = mockFetch.mock.calls[0];
      expect(url).toContain('/bookings/bk-1/ics');
    });

    it('P: calendar-link returns google url', async () => {
      const link = { googleCalendarUrl: 'https://calendar.google.com/calendar/render?action=TEMPLATE&text=DEKAT', icsUrl: '/api/v1/bookings/bk-1/ics', icsContent: 'BEGIN:VCALENDAR' };
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: link }));
      const res = await publicApi.bookings.calendarLink('bk-1');
      expect(res.data.googleCalendarUrl).toContain('calendar.google.com');
      expect(res.data.icsUrl).toContain('/ics');
    });

    it('A: ics with special chars still VCALENDAR', async () => {
      const ics = 'BEGIN:VCALENDAR\r\nSUMMARY:DEKAT Booking DKT\\,TEST\\;2\r\nEND:VCALENDAR';
      mockFetch.mockResolvedValueOnce({ ok: true, status: 200, text: () => Promise.resolve(ics), json: () => Promise.resolve(ics) } as unknown as Response);
      const res = await publicApi.bookings.ics('bk-special');
      expect(mockFetch).toHaveBeenCalled();
    });
  });

  describe('FAQ & Policy', () => {
    it('P: GET /public/faqs returns list', async () => {
      const faqs = [{ id: 'f1', question: 'Bagaimana cara booking?', answer: 'Pilih provider...' }];
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: faqs }));
      const res = await publicApi.faqs.listPublic();
      expect(res.data[0].question).toContain('booking');
    });

    it('P: GET /public/policies returns cancellation policy', async () => {
      const pols = [{ id: 'p1', title: 'Kebijakan Pembatalan', body: '24h', type: 'cancellation' }];
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: pols }));
      const res = await publicApi.policies.listPublic();
      expect(res.data[0].type).toBe('cancellation');
    });

    it('N: POST /provider/faqs without auth 401', async () => {
      mockFetch.mockResolvedValueOnce(mockResponse({ success: false, message: 'Unauthorized' }, false, 401));
      await expect(publicApi.faqs.createProvider({ question: 'Q', answer: 'A' })).rejects.toThrow();
    });

    it('P: POST /provider/faqs success', async () => {
      const created = { id: 'f2', question: 'Q2', answer: 'A2' };
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: created }));
      const res = await publicApi.faqs.createProvider({ question: 'Q2', answer: 'A2', category: 'booking' });
      expect(res.data.id).toBe('f2');
    });

    it('P: GET /provider/faqs list after create', async () => {
      mockFetch.mockResolvedValueOnce(mockResponse({ success: true, data: [{ id: 'f1' }, { id: 'f2' }] }));
      const res = await publicApi.faqs.listProvider();
      expect(res.data).toHaveLength(2);
    });
  });
});
